web/client/components/map/openlayers/plugins/WMSLayer.js

```diff
@@ -154,7 +154,50 @@ const createLayer = (options, map) => {
                 crossOrigin: options.crossOrigin,
                 attributions: toOLAttributions(options.credits),
                 params: queryParameters,
-                ratio: options.ratio || 1
+                ratio: options.ratio || 1,
+                imageLoadFunction: function(image, src) {
+                    // fixes #3916, see https://gis.stackexchange.com/questions/175057/openlayers-3-wms-styling-using-sld-body-and-post-request
+                    var img = image.getImage();
+
+                    if (typeof window.btoa === 'function' && src.length >= 2000) {
+                        const xhr = new XMLHttpRequest();
+                        // GET ALL THE PARAMETERS OUT OF THE SOURCE URL**
+                        const dataEntries = src.split("&");
+                        let url;
+                        let params = "";
+                        for (let i = 0; i < dataEntries.length; i++) {
+                            if (i === 0) {
+                                url = dataEntries[i];
+                            } else {
+                                params = params + "&" + dataEntries[i];
+                            }
+                        }
+                        xhr.open('POST', url, true);
+
+                        xhr.responseType = 'arraybuffer';
+                        xhr.onload = function() {
```

Comment:
Here we are using XHR object directly, this could fall into same origin policy violations, that normal images do not have.
@mbarto what do you think we should do here?
We could use axios, but we should check how it works with authentication (e.g. authkey, it is applied as additional parameter)